rig-core/src/agent/prompt_request.rs

```diff
@@ -0,0 +1,231 @@
+use std::{
+    future::{Future, IntoFuture},
+    marker::PhantomData,
+};
+
+use futures::{future::BoxFuture, stream, FutureExt, StreamExt};
+
+use crate::{
+    completion::{Completion, CompletionError, CompletionModel, Message, PromptError},
+    message::{AssistantContent, UserContent},
+    tool::ToolSetError,
+    OneOrMany,
+};
+
+use super::Agent;
+
+pub trait State {}
+pub struct Simple;
+pub struct MultiTurn;
+
+impl State for Simple {}
+impl State for MultiTurn {}
+
+pub trait SendPromptRequest<M: CompletionModel, T: State> {
+    fn send(self) -> impl Future<Output = Result<String, PromptError>> + Send;
+}
+
+/// A builder for creating prompt requests with customizable options.
+/// Uses generics to track which options have been set during the build process.
+pub struct PromptRequest<'c, 'a, M: CompletionModel, T: State> {
+    /// The prompt message to send to the model
+    prompt: Message,
+    /// Optional chat history to include with the prompt
+    /// Note: chat history needs to outlive the agent as it might be used with other agents
+    chat_history: Option<&'c mut Vec<Message>>,
+    /// Maximum depth for multi-turn conversations (0 means no multi-turn)
+    max_depth: usize,
+    /// The agent to use for execution
+    agent: &'a Agent<M>,
+
+    /// Typestate
+    _state: PhantomData<T>,
+}
```

Comment:
You don't need two lifetimes here, both `chat_history` and `agent` can share the same one without any issues (at least from my limited testing).

Looking at this StackOverflow [discussion](https://stackoverflow.com/questions/29861388/when-is-it-useful-to-define-multiple-lifetimes-in-a-struct), I don't think we fall in a case where this would cause a problem

Reply:
Yea, I think you are correct. I was trying to ensure that `chat_history` can exist longer than `agent` but i presume that relation isn't relevant in the context of `PromptRequest`